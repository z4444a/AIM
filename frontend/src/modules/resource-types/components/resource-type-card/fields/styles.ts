import createStyles from '@material-ui/core/styles/createStyles';
import { Theme, WithStyles } from '@material-ui/core';

export const styles = (theme: Theme) =>
  createStyles({
    mainContainer: {
      '& > div': {
        marginTop: theme.spacing.unit * 2,
      },
    },
    imgContainer: {
      display: 'flex',
    },
    avatar: {
      marginRight: theme.spacing.unit,
    },
    checkbox: {
      display: 'inline-block',
    },
  });

export type Styles = WithStyles<'mainContainer' | 'avatar' | 'imgContainer' | 'checkbox'>;
