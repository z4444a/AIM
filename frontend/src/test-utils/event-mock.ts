export const getMockChangeEvent = (value: string) => {
  return {
    target: {
      value: value,
    },
  };
};

export const getMockKeyPressEvent = (key: string) => {
  return {
    key: key,
    preventDefault: jest.fn().mockName('preventDefault'),
  };
};
