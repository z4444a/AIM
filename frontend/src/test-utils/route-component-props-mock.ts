import { RouteComponentProps } from 'react-router';
import { Href } from 'history';

export function getMockRouterProps<P>(data: P): RouteComponentProps {
  const location = {
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: {},
  };

  const props: RouteComponentProps<P> = {
    match: {
      isExact: true,
      params: data,
      path: '',
      url: '',
    },
    location: location,
    history: {
      length: 2,
      action: 'POP',
      location: location,
      push: () => {},
      replace: () => {},
      go: () => {},
      goBack: () => {},
      goForward: () => {},
      block: () => () => {},
      createHref: () => {
        const temp: Href = '';
        return temp;
      },
      listen: () => () => {},
    },
    staticContext: {},
  };

  return props;
}
