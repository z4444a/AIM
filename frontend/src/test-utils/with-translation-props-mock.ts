import { WithTranslation } from 'react-i18next';

export function getMockWithTranslationProps(): WithTranslation {
  // eslint-disable-next-line
  const t = ((key: string) => `${key}_i18n`) as any;
  // eslint-disable-next-line
  const i18n = {} as any;
  const props: WithTranslation = {
    tReady: true,
    t,
    i18n,
  };

  return props;
}
