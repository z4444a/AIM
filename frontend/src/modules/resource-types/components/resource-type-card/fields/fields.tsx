import React from 'react';
import { FormControlLabel } from '@material-ui/core';
import TextField from '@material-ui/core/TextField';
import { WithTranslation } from 'react-i18next';
import Checkbox from '@material-ui/core/Checkbox';
import { Styles } from './styles';
import Avatar from '@material-ui/core/Avatar';
import Tooltip from '@material-ui/core/Tooltip';
import Button from '@material-ui/core/Button';
import { ErrorState } from '../card';

export interface Props {
  errors: ErrorState;
  imgBytes: number[];
  imgSource: string;
  name: string;
  uniqueName: boolean;
  description: string;
  active: boolean;
  quantitative: boolean;
  backup: boolean;

  onImgByteCodeChange: (bytes: number[]) => void;
  onImgSourceChange: React.ChangeEventHandler<HTMLTextAreaElement>;
  onNameChange: React.ChangeEventHandler<HTMLTextAreaElement>;
  onDescriptionChange: React.ChangeEventHandler<HTMLTextAreaElement>;
  onActiveChange: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  onBackupChange: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;
  onQuantitativeChange: (event: React.ChangeEvent<HTMLInputElement>, checked: boolean) => void;

  highlightEmpty: boolean;
}

export type InternalProps = Props & WithTranslation & Styles;
export class TypeFields extends React.PureComponent<InternalProps> {
  public render(): React.ReactNode {
    const {
      classes,
      imgSource,
      name,
      uniqueName,
      description,
      active,
      backup,
      quantitative,
      onImgSourceChange,
      onNameChange,
      onDescriptionChange,
      onActiveChange,
      onBackupChange,
      onQuantitativeChange,
      highlightEmpty,
      t,
      errors,
    } = this.props;
    const src = this.getImageInBase64();
    return (
      <div className={classes.mainContainer}>
        <div className={classes.imgContainer}>
          <Avatar src={src} className={classes.avatar} />
          <TextField
            fullWidth
            value={imgSource || ''}
            onChange={onImgSourceChange}
            label={t('typeCard.fields.icon.label')}
            error={errors && !!errors.picture}
            helperText={errors && errors.picture}
          />
          <input
            accept="*"
            style={{ display: 'none' }}
            id="raised-button-file"
            type="file"
            multiple
            onChange={e => this.upload(e.target.files)}
          />
          <label htmlFor="raised-button-file">
            <Button variant="contained" component="span">
              {t('common.upload')}
            </Button>
          </label>
        </div>
        <div>
          <TextField
            fullWidth
            required
            error={!uniqueName || (highlightEmpty && !name.length)}
            helperText={!uniqueName ? t('typeCard.error.uniqueTypeName') : undefined}
            label={t('typeCard.fields.nameLabel')}
            value={name || ''}
            onChange={onNameChange}
          />
        </div>
        <div>
          <div className={classes.checkbox}>
            <Tooltip title={t(`typeCard.tooltips.active.${active}`)}>
              <FormControlLabel
                control={<Checkbox checked={active || false} onChange={onActiveChange} />}
                label={t('typeCard.fields.activeLabel')}
              />
            </Tooltip>
          </div>
          <div className={classes.checkbox}>
            <FormControlLabel
              control={<Checkbox checked={backup || false} onChange={onBackupChange} />}
              label={t('typeCard.fields.backupLabel')}
            />
          </div>
          <div className={classes.checkbox}>
            <Tooltip title={t(`typeCard.tooltips.quantitative.${quantitative}`)}>
              <FormControlLabel
                control={
                  <Checkbox checked={quantitative || false} onChange={onQuantitativeChange} />
                }
                label={t('typeCard.fields.quantitativeLabel')}
              />
            </Tooltip>
          </div>
        </div>
        <div>
          <TextField
            fullWidth
            label={t('typeCard.fields.descriptionLabel')}
            value={description || ''}
            onChange={onDescriptionChange}
            rows={6}
            multiline
          />
        </div>
      </div>
    );
  }

  private getImageInBase64 = () => {
    const { imgSource, imgBytes } = this.props;
    if (imgSource || (imgBytes && imgBytes.length === 0)) {
      return imgSource;
    }
    const image: string =
      typeof imgBytes === 'string'
        ? imgBytes
        : btoa(String.fromCharCode.apply(null, Array.from(new Uint8Array(imgBytes))));
    return image.length === 0 ? image : 'data:image/png;base64,' + image;
  };
  private upload = (files: FileList | null) => {
    const { onImgByteCodeChange } = this.props;
    if (files === null || files.length !== 1) {
      return;
    }
    const file = files[0];
    const reader = new FileReader();
    reader.onload = function() {
      const arrayBuffer = reader.result;
      if (typeof arrayBuffer === 'string' || arrayBuffer === null) {
        return;
      }
      const bytes = new Int8Array(arrayBuffer);
      onImgByteCodeChange(Array.from(bytes));
    };
    reader.readAsArrayBuffer(file);
  };
}
