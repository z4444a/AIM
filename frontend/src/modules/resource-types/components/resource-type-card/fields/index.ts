import { withTranslation } from 'react-i18next';
import { TypeFields } from './fields';
import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './styles';

export default withTranslation('common')(withStyles(styles)(TypeFields));
