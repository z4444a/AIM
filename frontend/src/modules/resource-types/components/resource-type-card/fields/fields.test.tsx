import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { InternalProps, TypeFields } from './fields';
import { getMockWithTranslationProps } from '../../../../../test-utils/with-translation-props-mock';

describe('ResourceTypeFields', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const props: InternalProps = {
      ...getMockWithTranslationProps(),
      classes: {
        avatar: 'avatar',
        imgContainer: 'imgContainer',
        mainContainer: 'mainContainer',
      },
      active: false,
      backup: false,
      description: '',
      highlightEmpty: false,
      imgSource: '',
      name: '',
      onActiveChange: (e, checked) => null,
      onBackupChange: (event, checked) => null,
      onDescriptionChange: event1 => null,
      onImgSourceChange: event1 => null,
      onNameChange: event1 => null,
    };
    wrapper = shallow(<TypeFields {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
