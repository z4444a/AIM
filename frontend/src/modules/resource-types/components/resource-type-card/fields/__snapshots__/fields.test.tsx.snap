// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTypeFields should render correctly 1`] = `
<div
  className="mainContainer"
>
  <div
    className="imgContainer"
  >
    <WithStyles(Avatar)
      className="avatar"
      src=""
    />
    <TextField
      fullWidth={true}
      label="typeCard.fields.icon.label_i18n"
      onChange={[Function]}
      required={false}
      select={false}
      value=""
      variant="standard"
    />
    <input
      accept="*"
      id="raised-button-file"
      multiple={true}
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <label
      htmlFor="raised-button-file"
    >
      <WithStyles(Button)
        component="span"
        variant="contained"
      >
        common.upload_i18n
      </WithStyles(Button)>
    </label>
  </div>
  <div>
    <TextField
      error={true}
      fullWidth={true}
      helperText="typeCard.error.uniqueTypeName_i18n"
      label="typeCard.fields.nameLabel_i18n"
      onChange={[Function]}
      required={true}
      select={false}
      value=""
      variant="standard"
    />
  </div>
  <div>
    <div>
      <WithStyles(Tooltip)
        title="typeCard.tooltips.active.false_i18n"
      >
        <WithStyles(WithFormControlContext(FormControlLabel))
          control={
            <WithStyles(Checkbox)
              checked={false}
              onChange={[Function]}
            />
          }
          label="typeCard.fields.activeLabel_i18n"
        />
      </WithStyles(Tooltip)>
    </div>
    <div>
      <WithStyles(WithFormControlContext(FormControlLabel))
        control={
          <WithStyles(Checkbox)
            checked={false}
            onChange={[Function]}
          />
        }
        label="typeCard.fields.backupLabel_i18n"
      />
    </div>
    <div>
      <WithStyles(Tooltip)
        title="typeCard.tooltips.quantitative.undefined_i18n"
      >
        <WithStyles(WithFormControlContext(FormControlLabel))
          control={
            <WithStyles(Checkbox)
              checked={false}
            />
          }
          label="typeCard.fields.quantitativeLabel_i18n"
        />
      </WithStyles(Tooltip)>
    </div>
  </div>
  <div>
    <TextField
      fullWidth={true}
      label="typeCard.fields.descriptionLabel_i18n"
      multiline={true}
      onChange={[Function]}
      required={false}
      rows={6}
      select={false}
      value=""
      variant="standard"
    />
  </div>
</div>
`;
